### PR TITLE
Add UTXO hash component in making pre-images

### LINKED ISSUE #849: Discussion: Should a node be able to enroll multiple times with the same address ?
Currently, our pre-image scheme relies only on the private key, a nonce, and a fixed string.
This leads to a situation where a user would get the same pre-image if they enrolled with two different frozen txs controlled by the same private key. This might not be wanted, or this could be a feature. It's not an issue with the protocol, but rather our implementation. Thoughts ?

CC @bpfkorea/core

### PR DESCRIPTION
**Pull request by linked0**
Our previous pre-image scheme relied only on the private key, a nonce, and a fixed string. But the scheme made it unable for a node to enroll with different frozen UTXOs because it made the same pre-images for the UTXOs.  So we have made it possible to enroll with different UTXOs by adding UXTO hash to our pre-image scheme.

Fixes #849

### CODE REVIEW

File: source/agora/consensus/EnrollmentManager.d [hunk 1]
```diff
@@ -660,6 +666,58 @@ public class EnrollmentManager
...
+
+            if (!results.empty)
+                en_key = results.oneValue!(ubyte[]).deserializeFull!(Hash);
+        }
+        catch (Exception ex)
+        {
+            log.error("ManagedDatabase operation error: {}", ex);
+        }
+
+        return en_key;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: A slight simplification to make it obvious when `Hash.init` is returned.

```suggestion
        try
        {
            auto results = this.db.execute("SELECT val FROM node_enroll_data " ~
                "WHERE key = ?", "enroll_key");

            if (!results.empty)
                return results.oneValue!(ubyte[]).deserializeFull!(Hash);
        }
        catch (Exception ex)
        {
            log.error("ManagedDatabase operation error: {}", ex);
        }

        return Hash.init;
```

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
LGTM, just one thing.

**Comment by linked0** [comment]
I applied Drey's requests. Thanks.
But CI builds have some errors on all the `ldc-master` environments about `ocean` as follows. If anyone knows how to deal with it, please let me know.

```
/home/runner/work/agora/agora/submodules/ocean/src/ocean/core/array/Search.d:73: error: undefined reference to '_D5ocean4core5array17DefaultPredicatesQt__T7IsEqualTaZQl6opCallFNaNbNiNfIaIaZb'
/home/runner/work/agora/agora/submodules/ocean/src/ocean/core/array/Transformation.d:71: error: undefined reference to '_D5ocean4core5array17DefaultPredicatesQt__T6IsLessTaZQk6opCallFNaNbNiNfIaIaZb'
/home/runner/work/agora/agora/submodules/ocean/src/ocean/core/array/Transformation.d:73: error: undefined reference to '_D5ocean4core5array17DefaultPredicatesQt__T6IsLessTaZQk6opCallFNaNbNiNfIaIaZb'
/home/runner/work/agora/agora/submodules/ocean/src/ocean/core/array/Transformation.d:123: error: undefined reference to '_D5ocean4core5array17DefaultPredicatesQt__T6IsLessTaZQk6opCallFNaNbNiNfIaIaZb'
```

**Comment by AndrejMitrovic** [comment]
It's not a bug with your PR. It seems latest ldc-master is broken: https://github.com/bpfkorea/agora/runs/1279452560

**Comment by linked0** [comment]
You can review my PR, all the builds on ldc-master have failed though.

File: source/agora/consensus/EnrollmentManager.d [hunk 2]
```diff
@@ -160,6 +160,9 @@ public class EnrollmentManager
         // create Pair object from KeyPair object
         this.key_pair = Pair.fromScalar(secretKeyToCurveScalar(key_pair.secret));
 
+        // load enrollment key
+        this.enroll_key = this.getEnrollmentKey();
+
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: This value is never read. Currently a validator wishing to enroll will call the enrollment manager with the UTXO key. Do we need to read it here ?

File: source/agora/consensus/EnrollmentManager.d [hunk 3]
```diff
@@ -168,8 +168,11 @@ public class EnrollmentManager
 
         // load the count of 'populate' of `PreImageCycle`
         const uint populate_count = this.getCycleIndex();
+
+        // if there exists an erollment key, `populate_count` must be above 0.
+        assert(this.enroll_key == Hash.init || populate_count > 0);
         foreach (_; 0 .. populate_count)
-            this.cycle.populate(this.key_pair.v, true);
+            this.cycle.populate(this.key_pair.v, this.enroll_key, true);
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: I see now why you are loading it in the  constructor. I think we should rather move this outside of the constructor, so pre-images only get generated when needed (and possible). Otherwise the first time a validator runs, tons of pre-images will be generated for nothing.

File: source/agora/consensus/PreImage.d [hunk 4]
```diff
@@ -351,14 +352,14 @@ public struct PreImageCycle
 
     ***************************************************************************/
 
-    public Hash populate (scope const ref Scalar secret, bool consume)
-        @safe nothrow
+    public Hash populate (scope const ref Scalar secret,
+        scope const ref Hash enroll_key, bool consume) @safe nothrow
```
Comment (Geod24) [inline source/agora/consensus/PreImage.d:356]: Adding the `enroll_key` changes quite a bit. When we do see another validator with the same private key, we don't want to override our own pre-images with their own.

**Review by Geod24 — CHANGES_REQUESTED** [review]
This needs a network test with two nodes.

File: source/agora/consensus/EnrollmentManager.d (hunk 3 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d]: It would be better to make pre-images when needed. I will find the right place to do it.

**Comment by linked0** [comment]
I added a network test code in the module, agora.test.PreImage. But, it's not the final code. I just uploaded the code for sharing what I'm trying to do.
@Geod24 I'd like you to check if it's the right approach. In the code, the `SameKeyValidator` is using the `WK.Keys.A` and the second UTXO for the key in order to make an Enrollment.

**Comment by linked0** [comment]
I'm writing a network test in this PR for a new validator with the same address as the existing validator. But, I'm not sure it's the right approach. Some whiteboarding session is needed I guess. With the last commit about network test, I have some assertion failure as follows. I'm debugging it now.

```
(lldb) bt
error: need to add support for DW_TAG_base_type 'char' encoded with DW_ATE = 0x10, bit_size = 8
error: need to add support for DW_TAG_base_type 'immutable(char)' encoded with DW_ATE = 0x10, bit_size = 8
* thread #64, stop reason = signal SIGABRT
  * frame #0: 0x00007fff6a52d33a libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff6a5e9e60 libsystem_pthread.dylib`pthread_kill + 430
    frame #2: 0x00007fff6a4b4808 libsystem_c.dylib`abort + 120
    frame #3: 0x00007fff67714458 libc++abi.dylib`abort_message + 231
    frame #4: 0x00007fff677058a7 libc++abi.dylib`demangling_terminate_handler() + 238
    frame #5: 0x00007fff692405b1 libobjc.A.dylib`_objc_terminate() + 104
    frame #6: 0x00007fff67713887 libc++abi.dylib`std::__terminate(void (*)()) + 8
    frame #7: 0x00007fff67716387 libc++abi.dylib`__cxa_rethrow + 99
    frame #8: 0x000000010073f0ac agora-unittests`stellar::Slot::processEnvelope(this=0x000000010c7182d8, envelope=0x0000000240412d30, self=false) at Slot.cpp:155:9
    frame #9: 0x000000010072e4a0 agora-unittests`stellar::SCP::receiveEnvelope(this=0x0000000104734aa0, envelope=0x0000000240412d30) at SCP.cpp:33:38
    frame #10: 0x0000000100142933 agora-unittests`_D5agora9consensus8protocol9NominatorQk15receiveEnvelopeMFNeMKxS4scpd5types11Stellar_SCP11SCPEnvelopeZv(this=0x00000001d901f2c0, envelope=0x0000000240412d30) at Nominator.d:444:9
    frame #11: 0x00000001001443a1 agora-unittests`_D5agora4node9ValidatorQk15receiveEnvelopeMFNfS4scpd5types11Stellar_SCP11SCPEnvelopeZv(this=0x000070000608a648, envelope=SCPEnvelope @ 0x0000000240412d30) at Validator.d:266:9
    frame #12: 0x000000010019670f agora-unittests`_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPITSQBaQxQu10SerializerZQCg13handleCommandFSQDuQDq7CommandQDdSQEmQEi9FilterAPIZv(cmd=Command @ 0x0000000240413ec0, node=0x000070000608a900, filter=FilterAPI @ 0x0000000240413ea0) at LocalRest.d-mixin-477:494:33
    frame #13: 0x00000001001aab53 agora-unittests`_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPITSQBaQxQu10SerializerZQCg__T7spawnedTCQCkQChQCf17TestValidatorNodeZQBnFNbSQFa11concurrency3TidAyaiSQEt6common6ConfigQhPOSQGv8RegistryQjAySQGg9consensus4data5BlockQgmPOlZ7runNodeMFZ__T6handleTSQJoQJk7CommandZQyMFQuZ9__lambda2MFZv at LocalRest.d:602:43
    frame #14: 0x00000001007e58f9 agora-unittests`_D6geod2411concurrency14FiberScheduler6createMFNbDFZvZ4wrapMFZv at concurrency.d:915:13
    frame #15: 0x0000000100d4551f agora-unittests`_D4core6thread7context8Callable6opCallMFZv(this=0x0000000000000002) at context.d:46:17
    frame #16: 0x0000000100d456ae agora-unittests`_D4core6thread5fiber5Fiber3runMFZv(this=0x00000001d951e680) at fiber.d:1129:9
    frame #17: 0x0000000100d455e3 agora-unittests`fiber_entryPoint at fiber.d:186:13
```

**Comment by linked0** [comment]
Ready for review.
The network test in the last commit should have been more complete. But, [the problem](https://github.com/bpfkorea/agora/pull/1268#issuecomment-716384194) could not be easily solved. If a validator re-enroll with another UTXO before its current enrollment expires, the problem occurs. So it would be good to handle it as a separate issue because it would take some time for me to analyze and get better knowledge about that. I'll create another issue for that if possible.

File: source/agora/consensus/EnrollmentManager.d [hunk 5]
```diff
@@ -1367,3 +1432,20 @@ unittest
...
+    auto utxo_set = new TestUTXOSet;
+    genesisSpendable()
+        .map!(txb => txb.refund(WK.Keys.A.address).sign(TxType.Freeze))
+        .each!(tx => utxo_set.put(tx));
+    auto man = new EnrollmentManager(":memory:", WK.Keys.A,
+        new immutable(ConsensusParams)(10));
+
+    man.setEnrollmentKey(utxo_set.keys[0]);
+    assert(man.getEnrollmentKey()[] == utxo_set.keys[0][]);
+    assert(man.getEnrollmentKey()[] != utxo_set.keys[1][]);
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: I would add an `assert` on the length of the UTXO set (should be 8) before L1448, and use a different value (e.g. `keys[3]`), so you make sure it doesn't pick the first one in the `keys` for whatever reason.

File: source/agora/consensus/EnrollmentManager.d [hunk 6]
```diff
@@ -664,6 +667,68 @@ public class EnrollmentManager
         return false;
     }
 
+    /***************************************************************************
+
+        Get the key for the enrollment for this node
+
+        Returns:
+            the key for the enrollment
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Note that this function can "fail silently". No exception being thrown, and only a log message.
Originally I thought it was a bug, but it makes sense given the current design of the `EnrollmentManager`.
If that's intentional, it should be documented, and also the `log.error` should be downgraded to a non-error.

File: source/agora/consensus/EnrollmentManager.d [hunk 7]
```diff
@@ -664,6 +667,68 @@ public class EnrollmentManager
...
+
+        return Hash.init;
+    }
+
+    /***************************************************************************
+
+        Set the key for the enrollment for this node
+
+        Params:
+            enroll_key = the key for the enrollment
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Ditto about failure documentation and logging.

File: source/agora/consensus/EnrollmentManager.d [hunk 8]
```diff
@@ -871,19 +799,22 @@ public class EnrollmentManager
         Params:
             height = block height to check
 
+        Returns:
+            The distance
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Or `uint.max` if we're not enrolled

File: source/agora/consensus/EnrollmentManager.d [hunk 9]
```diff
@@ -278,15 +281,24 @@ public class EnrollmentManager
...
-                (this.cycle.nonce * this.cycle.NumberOfCycles) + this.cycle.index);
+                // save the count of 'populate' of `PreImageCycle`
+                this.setCycleIndex(
+                    (this.cycle.nonce * this.cycle.NumberOfCycles)
+                        + this.cycle.index);
+            }
+            else
+            {
+                log.warn("New enrollment appears before the current one is " ~
+                    "expired, New: {}, Old: {}", enroll.utxo_key, this.enroll_key);
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Rephrasing: "Seen enrollment for a validator with the same private key! New enrollment: {}, current one: {}".
Adding block height would be even better, if possible:
"Seen enrollment for a validator with the same private key at block {}! New enrollment: {}, current one: {} (since #{})"

File: source/agora/consensus/EnrollmentManager.d [hunk 10]
```diff
@@ -278,15 +281,24 @@ public class EnrollmentManager
...
 
-            // save the count of 'populate' of `PreImageCycle`
-            this.setCycleIndex(
-                (this.cycle.nonce * this.cycle.NumberOfCycles) + this.cycle.index);
+                // save the count of 'populate' of `PreImageCycle`
+                this.setCycleIndex(
+                    (this.cycle.nonce * this.cycle.NumberOfCycles)
+                        + this.cycle.index);
+            }
+            else
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: If you move it as the first check and `return` you can simplify the code / diff.

File: source/agora/consensus/EnrollmentManager.d [hunk 11]
```diff
@@ -309,7 +321,7 @@ public class EnrollmentManager
     public Enrollment createEnrollment (Hash utxo) @safe nothrow
     {
         // K, frozen UTXO hash
-        this.enroll_key = utxo;
+        // this.enroll_key = utxo;
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Remove then ?

File: source/agora/test/SameAddressPreImage.d [hunk 12]
```diff
@@ -0,0 +1,198 @@
+module agora.test.SameAddressPreImage;
```
Comment (Geod24) [inline source/agora/test/SameAddressPreImage.d]: DDOC

**Review by Geod24 — COMMENTED** [review]
Nice. Looks like this is almost ready. The network test usage of shared leaves me a bit skeptical, and might be the cause of your issues. Let's look at it together tomorrow.

**Comment by linked0** [comment]
> Nice. Looks like this is almost ready. The network test usage of shared leaves me a bit skeptical and might be the cause of your issues. Let's look at it together tomorrow.

Thank you for the review. I'll fix what you commented on.
In fact, I didn't add the code in the last commit that makes the problem happen. But I will add the code as another network test with a  `version (none)` modifier.

**Comment by linked0** [comment]
Ready for review after applying reviewers' comments.
I also added the network test code that makes [the problem](https://github.com/bpfkorea/agora/pull/1268#issuecomment-716384194) happen with the `version(none)` modifier.

File: source/agora/consensus/EnrollmentManager.d [hunk 13]
```diff
@@ -664,6 +667,77 @@ public class EnrollmentManager
         return false;
     }
 
+    /***************************************************************************
+
+        Get the key for the enrollment for this node
+
+        If getting an enrollment key fails, this displays a log messge and
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: message

File: source/agora/consensus/EnrollmentManager.d [hunk 14]
```diff
@@ -664,6 +667,77 @@ public class EnrollmentManager
...
+        returns `Hash.init`. If the database operation fails, it means that node
+        is not in the normal situation overall.
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: "the node is not in the normal situation" is not quite a good explanation. If we detect and error situation, we usually try to remedy it. In this code, I was assuming that the failure was "normal" in case the UTXO wasn't propagated yet. But that doesn't seem to be the case ?

Comment (Geod24) [reply source/agora/consensus/EnrollmentManager.d]: Looking at the code again, when the node is not enrolled, I assume the `SELECT` will return an empty result, so no exception will be thrown ? If so that should be documented.

File: source/agora/consensus/EnrollmentManager.d [hunk 15]
```diff
@@ -664,6 +667,77 @@ public class EnrollmentManager
...
+        }
+
+        return Hash.init;
+    }
+
+    /***************************************************************************
+
+        Set the key for the enrollment for this node
+
+        If saving an enrollment key fails, this displays a log messges and just
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Ditto typo

**Comment by Geod24** [comment]
Doesn't the third commit invalidate the test Genesis block ? It would seem so.
I extracted the first 2 commits here: https://github.com/bpfkorea/agora/pull/1303
Will look into the network test failure and the third commit.

**Comment by Geod24** [comment]
Partially rebased on v0.x.x, added the changes to the test genesis block, likely the reason your network test didn't pass.
I'll continue working on this.

**Comment by Geod24** [comment]
Okay, this actually uncovers TONS of problems. We generally identify validators by their public key, but since public keys are not unique, that creates issues. For example, the network registry relies on it.

I think we need to whiteboard (as a team) on this.

**Comment by AndrejMitrovic** [comment]
> but since public keys are not unique

You left us on a cliffhanger here. :D
